### ISSUE TITLE
Sharing parameters between models

### ISSUE THREAD

**Issue by bmazoure** [deepmind/dm-haiku#189]
I'm wondering whether there is a way to easily share/replace parameters of a model by another set of parameters.
For example, assume you have the following two models:
```
def _a(x):
            x = CNN(name='shared_encoder')(x)
            return hk.Linear(256,name='mlp_a')(x)

def _b(x):
            x = CNN(name='shared_encoder')(x)
            return hk.Linear(256,name='mlp_b')(x)
```

You then want to update `mlp_a,mlp_b` using their respective loss functions *but* share the encoder parameters between both. One way to do so is to filter out all `shared_encoder` parameters of b and replace them by the parameters of a, but this seems cumbersome and not straightforward.

Is there an easier way to go about solving this issue?

**Comment by tomhennigan** [comment]
Hi @bmazoure , in this specific case you might consider having your transformed function compute and return both a and b:

```python
def f(x):
  x = CNN(name='shared_encoder')(x)
  a = hk.Linear(256,name='mlp_a')(x)
  b = hk.Linear(256,name='mlp_a')(x)
  return a, b

f = hk.transform(f)

params = f.init(..)

def loss_fn(params, ..):
  a, b = f.apply(params, ..)
  a_loss = g(a, ..)
  b_loss = h(b, ..)
  return a_loss + b_loss

grad_fn = jax.jit(jax.grad(loss_fn))

grads = grad_fn(params, ..)
```

If you want separate apply functions for a and b then you can achieve that using [`hk.multi_transform`](https://dm-haiku.readthedocs.io/en/latest/api.html#multi-transform).
